storage/engine/cgo_flags.go

```diff
@@ -0,0 +1,15 @@
+package engine
+
+import (
+	// Link against the protobuf and rocksdb libaries. This is
+	// explicitly because these Go libraries do not export any Go
+	// symbols.
+	_ "github.com/cockroachdb/c-protobuf"
+	_ "github.com/cockroachdb/c-rocksdb"
+)
+
+// #cgo CXXFLAGS: -std=c++11
+// #cgo CPPFLAGS: -I ../../../c-protobuf/internal/src -I ../../../c-rocksdb/internal/include
+// #cgo darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup
```

Comment:
The problem is here, not in the upstream packages. The issue is with how "go build" links the compiled objects into a single object and does not tolerate undefined symbols. Frankly, it seems like a bug in cgo. Previously this wasn't a problem because the `-lrocksdb`, `-lprotobuf`, etc flags are part of the link line and those libraries contain the undefined symbols. 

You are correct that this is not darwin vs !darwin but clang vs gcc. I'm not aware of a build tag which allows you to distinguish clang vs gcc. I'll add a comment explaining this.